### PR TITLE
libnetwork/datastore: remove DataStore interface, and rename constructor

### PR DESCRIPTION
**Pull request by thaJeztah**
### libnetwork/datastore: move MockData to a _test file

It's only used in tests, and only within this package.

### libnetwork/datastore: remove DataStore interface

It only had a single implementation, so let's remove the interface.

While changing, also renaming;

- datastore.DataStore -> datastore.Store
- datastore.NewDataStore -> datastore.New
- datastore.NewDataStoreFromConfig -> datastore.FromConfig

### libnetwork/datastore: remove unused ParseKey() utility

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: libnetwork/controller.go [hunk 1]
```diff
@@ -341,7 +341,7 @@ func (c *Controller) makeDriverConfig(ntype string) map[string]interface{} {
 	}
 
 	if c.cfg.Scope.IsValid() {
-		// FIXME: every driver instance constructs a new DataStore
+		// FIXME: every driver instance constructs a new FoobarStore
```
Comment (thaJeztah) [inline libnetwork/controller.go:344]: Whoops. I used `FoobarStore` during the rename. I'll fix it back before merging 😂

**Review by corhere — APPROVED** [review]
LGTM!

I'm now tempted to try hitting datastore with a big generics stick to see if it might bring some semblance of sanity to the interface

**Comment by thaJeztah** [comment]
Yeah still peeling off layers one at a time; perhaps untangling the web still reveals more.

**Comment by corhere** [comment]
Welp, I tried to see if the List API could be changed to `func (ds *Store).List[T KVObject](key string) ([]T, error)`, then immediately hit a roadblock. There's at least one KVObject implementation with a dynamic `KeyPrefix()`, and `(*Store).List()` uses the object's `KeyPrefix()` ignoring the the `key` argument when `ds.cache != nil`, because of course the presence of a cache totally changes the semantics of the datastore in surprising ways.

**Comment by thaJeztah** [comment]
There's also one code-path I was eyeing; `datastore.List()` calls `datastore.iterateKVPairsFromStore`; https://github.com/moby/moby/blob/96b473a0bdc3f4c1cd0324f3ff044b04ab349df8/libnetwork/datastore/datastore.go#L347

Which calls `ds.ensureParent(key` https://github.com/moby/moby/blob/96b473a0bdc3f4c1cd0324f3ff044b04ab349df8/libnetwork/datastore/datastore.go#L361-L364

Which means we're _inserting_ a key if the datastore is empty? https://github.com/moby/moby/blob/96b473a0bdc3f4c1cd0324f3ff044b04ab349df8/libnetwork/datastore/datastore.go#L324-L333


Perhaps there was a reason for that, but I could consider that we forego that path if we're only _listing_ (and just return "nope, nothing there (yet)"

**Comment by thaJeztah** [comment]
@akerouanton LGTY?

File: libnetwork/datastore/datastore.go [hunk 2]
```diff
@@ -219,36 +197,37 @@ func NewDataStoreFromConfig(dsc discoverapi.DatastoreConfigData) (DataStore, err
...
 
-func (ds *datastore) Scope() string {
+// Scope returns the scope of the store.
+func (ds *Store) Scope() string {
 	return ds.scope
 }
 
-func (ds *datastore) KVStore() store.Store {
+// KVStore returns access to the KV Store.
+func (ds *Store) KVStore() store.Store {
```
Comment (akerouanton) [inline libnetwork/datastore/datastore.go:213]: This method is only used by a few tests. We can probably get rid of it.

**Review by akerouanton — APPROVED** [review]
LGTM; you can probably remove `KVStore()` in a subsequent PR.

File: libnetwork/datastore/datastore.go (hunk 2 above)
Comment (thaJeztah) [reply libnetwork/datastore/datastore.go:213]: Oh, good one; can probably take care of that in https://github.com/moby/moby/pull/46041, which is a follow-up to this one 👍

**Pull request merged**